File: code/ex112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sol_sum(self, root, sum, s, p):
        global mark
        
        if root != None:
            p = root
            
            s += root.val
            
            self.sol_sum(root.left, sum, s, p)
            if mark == 1: return
            self.sol_sum(root.right, sum, s, p)
            if mark == 1: return
        else:
            if (s == sum) and (p.left == None) and (p.right == None):
                mark = 1
                return
    
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        
        global mark
        mark = 0
        
        if root == None:
            return False
        
        self.sol_sum(root, sum, 0, root)
        
        if mark == 1:
            return True
        else:
            return False
        
        

File: code/ex112.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (root == NULL)    return 0;
        return helper(root, sum, 0);
    }
    
private:
    bool helper(TreeNode* root, int sum, int cur) {
        if (root->left == NULL && root->right == NULL)    return sum == cur + root->val;
        bool left = root->left == NULL ? 0 : helper(root->left, sum, cur + root->val);
        bool right = root->right == NULL ? 0 : helper(root->right, sum, cur + root->val);
        return left || right;
    }
};